File: api-tsl/csharp/src/util/TSLUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Xml;

namespace tr.gov.tubitak.uekae.esya.api.tsl.util
{
    public class TSLUtil
    {
        private static TSLUtil instance;
        private string _configPath = Application.StartupPath + "\\tsl-config.xml";

        public static TSLUtil getInstance()
        {
            if (instance == null)
            {
                instance = new TSLUtil();
            }
            return instance;
        }

        //public enum TSLTypeEnum
        //{
        //    Generic,
        //    Schemes
        //}

        //public enum StatusDetApproachEnum
        //{
        //    Active,
        //    Passive,
        //    Delinquent,
        //    List
        //}

        //public enum ServiceTypeIdEnum
        //{
        //    PKC,
        //    QC,
        //    TSA,
        //    OCSP,
        //    CRL,
        //    RA,
        //    IdV,
        //    CGen,
        //    ACA,
        //    Archiv,
        //    REM,
        //    KEscrow,
        //    PPwd,
        //    SignaturePolicyAuthority,
        //    Supervision,
        //    Voluntary,
        //    TSLIssuer,
        //    Unspecified
        //}

        //public enum ServiceCurrentStatusEnum
        //{
        //    InAccord,
        //    Expired,
        //    Suspended,
        //    Revoked,
        //    NotInAccord
        //}

        //public enum SchemeTypeCommRulesEnum
        //{
        //    Supervision,
        //    Volapproval
        //}

        private readonly XmlDocument mDocument = new XmlDocument();

        readonly IList<string> languageList = new List<string>();
        readonly IList<string> countryList = new List<string>();
        readonly IList<string> tslTypeList = new List<string>();
        readonly IList<string> statusDetApproachList = new List<string>();
        readonly IList<string> serviceTypeIdList = new List<string>();
        readonly IList<string> serviceCurrentStatusList = new List<string>();
        readonly IList<string> schemeTypeCommRulesList = new List<string>();

        public TSLUtil()
        {
            loadTSLConfigFile();
            initlanguageList();
            initCountryList();
        }
        public void setConfigPath(String path)
        {
            _configPath = path;
        }
        public String getConfigPath()
        {
            return _configPath;
        }
        private void loadTSLConfigFile()
        {
            mDocument.Load(getConfigPath());

            XmlNodeList nodeList = null;

            nodeList = mDocument.SelectNodes("tsl-config");
            if (nodeList != null && nodeList.Count == 1)
            {
                XmlElement rootElement = (XmlElement)nodeList[0];

                nodeList = null;

                nodeList = rootElement.SelectNodes("TSLType");
                if (nodeList != null && nodeList.Count == 1)
                {
                    XmlElement tslTypeElement = (XmlElement)nodeList[0];
                    initTSLTypeList(tslTypeElement);
                }
                nodeList = null;

                nodeList = rootElement.SelectNodes("StatusDeterminationApproach");
                if (nodeList != null && nodeList.Count == 1)
                {
                    XmlElement statusDetApprElement = (XmlElement)nodeList[0];
                    initStatusDetApproachList(statusDetApprElement);
                }
                nodeList = null;

                nodeList = rootElement.SelectNodes("ServiceTypeIdentifier");
                if (nodeList != null && nodeList.Count == 1)
                {
                    XmlElement serviceTypeIdElement = (XmlElement)nodeList[0];
                    initServiceTypeIdList(serviceTypeIdElement);
                }
                nodeList = null;

                nodeList = rootElement.SelectNodes("ServiceCurrentStatus");
                if (nodeList != null && nodeList.Count == 1)
                {
                    XmlElement serviceCurrentStatusElement = (XmlElement)nodeList[0];
                    initServiceCurrentStatusList(serviceCurrentStatusElement);
                }
                nodeList = null;

                nodeList = rootElement.SelectNodes("SchemeTypeCommunityRules");
                if (nodeList != null && nodeList.Count == 1)
                {
                    XmlElement schemeTypeCommRulesElement = (XmlElement)nodeList[0];
                    initSchemeTypeCommRulesList(schemeTypeCommRulesElement);
                }
                nodeList = null;
            }
            
        }

        private void initlanguageList()
        {
            languageList.Add("tr");
            languageList.Add("en");
            languageList.Add("bg");
            languageList.Add("cs");
            languageList.Add("da");
            languageList.Add("de");
            languageList.Add("el");
            languageList.Add("es");
            languageList.Add("et");
            languageList.Add("fi");
            languageList.Add("fr");
            languageList.Add("ga");
            languageList.Add("hu");
            languageList.Add("is");
            languageList.Add("it");
            languageList.Add("lb");
            languageList.Add("lt");
            languageList.Add("lv");
            languageList.Add("mt");
            languageList.Add("nb");
            languageList.Add("nl");
            languageList.Add("nn");
            languageList.Add("no");
            languageList.Add("pl");
            languageList.Add("pt");
            languageList.Add("ro");
            languageList.Add("sk");
            languageList.Add("sl");
            languageList.Add("sv");
        }

        public IList<string> LanguageList()
        {
            return languageList;
        }

        private void initCountryList()
        {
            countryList.Add("TR");
            countryList.Add("EU");
            countryList.Add("AT");
            countryList.Add("BE");
            countryList.Add("BG");
            countryList.Add("CY");
            countryList.Add("CZ");
            countryList.Add("DE");
            countryList.Add("DK");
            countryList.Add("EE");
            countryList.Add("EL");
            countryList.Add("ES");
            countryList.Add("FI");
            countryList.Add("FR");
            countryList.Add("HU");
            countryList.Add("IE");
            countryList.Add("IS");
            countryList.Add("IT");
            countryList.Add("LI");
            countryList.Add("LT");
            countryList.Add("LU");
            countryList.Add("LV");
            countryList.Add("MT");
            countryList.Add("NL");
            countryList.Add("NO");
            countryList.Add("PL");
            countryList.Add("PT");
            countryList.Add("RO");
            countryList.Add("SE");
            countryList.Add("SI");
            countryList.Add("SK");
            countryList.Add("UK");
        }

        public IList<string> CountryList()
        {
            return countryList;
        }

        private void initTSLTypeList(XmlElement iElement)
        {
            XmlNodeList nodeList = iElement.ChildNodes;
            foreach (var node in nodeList)
            {
                tslTypeList.Add(((XmlElement)node).InnerText);
            }
        }

        public IList<string> TSLTypeList()
        {
            return tslTypeList;
        }

        public string TSLTypeValue(int pos)
        {
            if (pos >= 0 && pos < tslTypeList.Count)
            {
                return tslTypeList[pos];
            }
            return null;
        }

        private void initStatusDetApproachList(XmlElement iElement)
        {
            XmlNodeList nodeList = iElement.ChildNodes;
            foreach (var node in nodeList)
            {
                statusDetApproachList.Add(((XmlElement)node).InnerText);
            }
        }

        public IList<string> StatusDeterminationApproach()
        {
            return statusDetApproachList;
        }

        public string StatusDeterminationApproachValue(int pos)
        {
            if (pos >= 0 && pos < statusDetApproachList.Count)
            {
                return statusDetApproachList[pos];
            }
            return null;
        }

        private void initServiceTypeIdList(XmlElement iElement)
        {
            XmlNodeList nodeList = iElement.ChildNodes;
            foreach (var node in nodeList)
            {
                serviceTypeIdList.Add(((XmlElement)node).InnerText);
            }
        }

        public IList<string> ServiceTypeId()
        {
            return serviceTypeIdList;
        }

        public string ServiceTypeIdValue(int pos)
        {
            if (pos >= 0 && pos < serviceTypeIdList.Count)
            {
                return serviceTypeIdList[pos];
            }
            return null;
        }

        private void initServiceCurrentStatusList(XmlElement iElement)
        {
            XmlNodeList nodeList = iElement.ChildNodes;
            foreach (var node in nodeList)
            {
                serviceCurrentStatusList.Add(((XmlElement)node).InnerText);
            }
        }

        public IList<string> ServiceCurrentStatus()
        {
            return serviceCurrentStatusList;
        }

        public string ServiceCurrentStatusValue(int pos)
        {
            if (pos >= 0 && pos < serviceCurrentStatusList.Count)
            {
                return serviceCurrentStatusList[pos];
            }
            return null;
        }

        private void initSchemeTypeCommRulesList(XmlElement iElement)
        {
            XmlNodeList nodeList = iElement.ChildNodes;
            foreach (var node in nodeList)
            {
                schemeTypeCommRulesList.Add(((XmlElement)node).InnerText);
            }
        }

        public IList<string> SchemeTypeCommunityRules()
        {
            return schemeTypeCommRulesList;
        }

        public string SchemeTypeComminityRulesValue(int pos)
        {
            if (pos >= 0 && pos < schemeTypeCommRulesList.Count)
            {
                return schemeTypeCommRulesList[pos];
            }
            return null;
        }

        public XmlNamespaceManager getNamespaceManager(XmlDocument iDocument)
        {
            XmlNameTable nsTable = iDocument.NameTable;
            XmlNamespaceManager nsManager = new XmlNamespaceManager(nsTable);
            nsManager.AddNamespace("tslx", Constants.NS_TSLX);
            nsManager.AddNamespace("tsl", Constants.NS_TSL);
            nsManager.AddNamespace("ds", Constants.NS_XMLDSIG);

            return nsManager;
        }

    }
}

File: api-tsl/src/tr/gov/tubitak/uekae/esya/api/tsl/util/TSLUtil.java
package tr.gov.tubitak.uekae.esya.api.tsl.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;
import tr.gov.tubitak.uekae.esya.api.xmlsignature.core.xml.NodeNamespaceContext;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static tr.gov.tubitak.uekae.esya.api.xmlsignature.core.Constants.NS_NAMESPACESPEC;

public class TSLUtil {
	protected static Logger logger = LoggerFactory.getLogger(TSLUtil.class);
	private static TSLUtil instance;
	private static String configPath="./config/tsl-config.xml";
	public static TSLUtil getInstance() {
		if (instance == null) {
			instance = new TSLUtil();
		}
		return instance;
	}

	private Document mDocument; // = new Document();

	List<String> languageList = new ArrayList<String>();
	List<String> countryList = new ArrayList<String>();
	List<String> tslTypeList = new ArrayList<String>();
	List<String> statusDetApproachList = new ArrayList<String>();
	List<String> serviceTypeIdList = new ArrayList<String>();
	List<String> serviceCurrentStatusList = new ArrayList<String>();
	List<String> schemeTypeCommRulesList = new ArrayList<String>();

	public TSLUtil() {
		try {
			loadTSLConfigFile();
		} catch (XPathExpressionException e) {
			logger.error("Error in TSLUtil", e);
		} catch (ParserConfigurationException e) {
			logger.error("Error in TSLUtil", e);
		} catch (SAXException e) {
			logger.error("Error in TSLUtil", e);
		} catch (IOException e) {
			logger.error("Error in TSLUtil", e);
		}
		initlanguageList();
		initCountryList();
	}
	public void setConfigPath(String path) {
		configPath=path;
	}
	public String getConfigPath() {
		return configPath;
	}
	private void loadTSLConfigFile() throws ParserConfigurationException,
			XPathExpressionException, SAXException, IOException {
		
		DocumentBuilder db = DocumentBuilderFactory.newInstance().newDocumentBuilder();
		//String path =TSL.class.getProtectionDomain().getCodeSource().getLocation().getPath();
		//String decodepath = "./config/tsl-config.xml";//URLDecoder.decode(path, "UTF-8");
		mDocument = db.parse(getConfigPath()); //+ "\\tsl-config.xml");
		// mDocument.Load(Application.StartupPath + "\\tsl-config.xml");
		
		Element rootElement=mDocument.getDocumentElement();
		
		NodeList nodeList = rootElement.getElementsByTagName("TSLType");		
		if (nodeList != null && nodeList.getLength() == 1) {
			Element tslTypeElement=(Element) nodeList.item(0);
			initTSLTypeList(tslTypeElement);
		}
		nodeList = null;
		
		nodeList = rootElement.getElementsByTagName("StatusDeterminationApproach");		
		if (nodeList != null && nodeList.getLength() == 1) {
			Element statusDetApprElement = (Element) nodeList.item(0);
			initStatusDetApproachList(statusDetApprElement);
		}
		nodeList = null;
		
		nodeList = rootElement.getElementsByTagName("ServiceTypeIdentifier");		
		if (nodeList != null && nodeList.getLength() == 1) {
			Element statusDetApprElement = (Element) nodeList.item(0);
			initStatusDetApproachList(statusDetApprElement);
		}
		nodeList = null;
		
		nodeList = rootElement.getElementsByTagName("ServiceCurrentStatus");		
		if (nodeList != null && nodeList.getLength() == 1) {
			Element serviceCurrentStatusElement = (Element) nodeList.item(0);
			initServiceCurrentStatusList(serviceCurrentStatusElement);
		}
		nodeList = null;
		
		nodeList = rootElement.getElementsByTagName("SchemeTypeCommunityRules");		
		if (nodeList != null && nodeList.getLength() == 1) {
			Element schemeTypeCommRulesElement = (Element) nodeList.item(0);
			initSchemeTypeCommRulesList(schemeTypeCommRulesElement);
		}
	}

	private void initlanguageList() {
		languageList.add("tr");
		languageList.add("en");
		languageList.add("bg");
		languageList.add("cs");
		languageList.add("da");
		languageList.add("de");
		languageList.add("el");
		languageList.add("es");
		languageList.add("et");
		languageList.add("fi");
		languageList.add("fr");
		languageList.add("ga");
		languageList.add("hu");
		languageList.add("is");
		languageList.add("it");
		languageList.add("lb");
		languageList.add("lt");
		languageList.add("lv");
		languageList.add("mt");
		languageList.add("nb");
		languageList.add("nl");
		languageList.add("nn");
		languageList.add("no");
		languageList.add("pl");
		languageList.add("pt");
		languageList.add("ro");
		languageList.add("sk");
		languageList.add("sl");
		languageList.add("sv");
	}

	public List<String> LanguageList() {
		return languageList;
	}

	private void initCountryList() {
		countryList.add("TR");
		countryList.add("EU");
		countryList.add("AT");
		countryList.add("BE");
		countryList.add("BG");
		countryList.add("CY");
		countryList.add("CZ");
		countryList.add("DE");
		countryList.add("DK");
		countryList.add("EE");
		countryList.add("EL");
		countryList.add("ES");
		countryList.add("FI");
		countryList.add("FR");
		countryList.add("HU");
		countryList.add("IE");
		countryList.add("IS");
		countryList.add("IT");
		countryList.add("LI");
		countryList.add("LT");
		countryList.add("LU");
		countryList.add("LV");
		countryList.add("MT");
		countryList.add("NL");
		countryList.add("NO");
		countryList.add("PL");
		countryList.add("PT");
		countryList.add("RO");
		countryList.add("SE");
		countryList.add("SI");
		countryList.add("SK");
		countryList.add("UK");
	}

	public List<String> CountryList() {
		return countryList;
	}

	private void initTSLTypeList(Element iElement) {
		NodeList nodeList = iElement.getElementsByTagName("*");
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			tslTypeList.add(((Element) node).getTextContent());
		}
	}

	public List<String> TSLTypeList() {
		return tslTypeList;
	}

	public String TSLTypeValue(int pos) {
		if (pos >= 0 && pos < tslTypeList.size()) {
			return tslTypeList.get(pos);
		}
		return null;
	}

	private void initStatusDetApproachList(Element iElement) {
		NodeList nodeList = iElement.getElementsByTagName("*");
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			statusDetApproachList.add(((Element) node).getTextContent());
		}
	}

	public List<String> StatusDeterminationApproach() {
		return statusDetApproachList;
	}

	public String StatusDeterminationApproachValue(int pos) {
		if (pos >= 0 && pos < statusDetApproachList.size()) {
			return statusDetApproachList.get(pos);
		}
		return null;
	}

	private void initServiceTypeIdList(Element iElement) {
		NodeList nodeList = iElement.getElementsByTagName("*");
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			serviceTypeIdList.add(((Element) node).getTextContent());
		}
	}

	public List<String> ServiceTypeId() {
		return serviceTypeIdList;
	}

	public String ServiceTypeIdValue(int pos) {
		if (pos >= 0 && pos < serviceTypeIdList.size()) {
			return serviceTypeIdList.get(pos);
		}
		return null;
	}

	private void initServiceCurrentStatusList(Element iElement) {
		NodeList nodeList = iElement.getElementsByTagName("*");
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			serviceCurrentStatusList.add(((Element) node).getTextContent());
		}
	}

	public List<String> ServiceCurrentStatus() {
		return serviceCurrentStatusList;
	}

	public String ServiceCurrentStatusValue(int pos) {
		if (pos >= 0 && pos < serviceCurrentStatusList.size()) {
			return serviceCurrentStatusList.get(pos);
		}
		return null;
	}

	private void initSchemeTypeCommRulesList(Element iElement) {
		NodeList nodeList = iElement.getElementsByTagName("*");
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			schemeTypeCommRulesList.add(((Element) node).getTextContent());
		}
	}

	public List<String> SchemeTypeCommunityRules() {
		return schemeTypeCommRulesList;
	}

	public String SchemeTypeComminityRulesValue(int pos) {
		if (pos >= 0 && pos < schemeTypeCommRulesList.size()) {
			return schemeTypeCommRulesList.get(pos);
		}
		return null;
	}

	public XPath getNamespaceManager(Document iDocument) {
		// not sure
		Element nscontext = XmlUtil.createDSctx(mDocument, "tsl",
				Constants.NS_TSL);

		XPathFactory factory = XPathFactory.newInstance();
		XPath xpath = factory.newXPath();
		xpath.setNamespaceContext(new NodeNamespaceContext(nscontext));
		return xpath;
	}
	public XPath getNamespaceManagerDS(Document iDocument) {
		// not sure
		Element nscontext = XmlUtil.createDSctx(mDocument, "ds",
				Constants.NS_XMLDSIG);

		XPathFactory factory = XPathFactory.newInstance();
		XPath xpath = factory.newXPath();
		xpath.setNamespaceContext(new NodeNamespaceContext(nscontext));

		return xpath;
	}
    public static Element createDSMultiple(Document doc, String prefix, String namespace)
    {
        Element ctx = doc.createElementNS(null, "namespaceContext");

        ctx.setAttributeNS(NS_NAMESPACESPEC, "xmlns:" + prefix.trim(), namespace);

        return ctx;
    }
}
